### PR TITLE
T26 - MyList <T>

### PR DESCRIPTION
**Pull request by Shresht07**

### CODE REVIEW

File: Training/Program.cs [hunk 1]
```diff
@@ -1,7 +1,77 @@
+using Training;
```
Comment (priyarathna) [inline Training/Program.cs:1]: Include comments correspondingly.

File: Training/Program.cs [hunk 2]
```diff
@@ -1,7 +1,77 @@
+using Training;
+
 namespace Training {
```
Comment (priyarathna) [inline Training/Program.cs:3]: 'Namespace' should be for the entire program.

File: Training/Program.cs [hunk 3]
```diff
@@ -1,7 +1,77 @@
+using Training;
+
 namespace Training {
-    internal class Program {
-        static void Main (string[] args) {
-            Console.WriteLine ("Hello, World!");
-        }
-    }
+   class MyList<T> {
+      T[] mVariables;
```
Comment (priyarathna) [inline Training/Program.cs:5]: Private variables should be declared at the end.

File: Training/Program.cs [hunk 4]
```diff
@@ -1,7 +1,77 @@
...
-    }
+   class MyList<T> {
+      T[] mVariables;
+      public MyList () => mVariables = new T[4];
+
+      public int Count => mSize;
+
+      public int Capacity => mVariables.Length;
+
+      public T this[int index] { get { return mVariables[index]; } set { mVariables[index] = value; } }
```
Comment (priyarathna) [inline Training/Program.cs:12]: Include the check to handle 'IndexOutOfRangeException'.

File: Training/Program.cs [hunk 5]
```diff
@@ -1,7 +1,77 @@
...
+
+      public void Add (T element) {
+         if (Count == Capacity) Array.Resize (ref mVariables, Capacity * 2);
+         mVariables[mSize++] = element;
+      }
+
+      public bool Remove (T element) {
+         if (mVariables.Contains (element)) {
```
Comment (priyarathna) [inline Training/Program.cs:20]: Store 'mVariables.Contains (element)' in a local var as it is used more than a time.

File: Training/Program.cs [hunk 6]
```diff
@@ -1,7 +1,77 @@
...
+            mVariables.Except (new T[] { element }).ToArray ();
```
Comment (priyarathna) [inline Training/Program.cs:21]: Array values are not stored.

File: Training/Program.cs [hunk 7]
```diff
@@ -1,7 +1,77 @@
...
+      list.Add (5);
+      list.Add (6);
+      PrintMyList (list);
+      Console.WriteLine ($"Count after adding elements in the list: {list.Count}");
+      Console.WriteLine ($"Capacity after adding elements: {list.Capacity}");
+      list.RemoveAt (0);
+      list.Insert (3, 3);
+      PrintMyList (list);
+      list.Clear ();
+      Console.Write ("After clearing list: ");
```
Comment (priyarathna) [inline Training/Program.cs:65]: Add an element after clearing the list. The output is wrong.

**Review by priyarathna — CHANGES_REQUESTED** [review]

File: Training/Program.cs [hunk 8]
```diff
@@ -1,7 +1,131 @@
...
+         list.Add (5);
+         list.Add (6);
+         PrintMyList (list);
+         Console.WriteLine ($"Count after adding elements in the list: {list.Count}");
+         Console.WriteLine ($"Capacity after adding elements: {list.Capacity}");
+         list.RemoveAt (0);
+         list.Insert (2, 3);
+         PrintMyList (list);
+         list.Clear ();
+         Console.Write ("After clearing list, list count => ");
```
Comment (priyarathna) [inline Training/Program.cs:116]: This is not the correct way to print the count. You can directly print the count.

File: Training/Program.cs [hunk 9]
```diff
@@ -1,7 +1,131 @@
...
+   internal class Program {
+      #region Methods -----------------------------------------------
+      /// <summary>Entry Point of the program</summary>
+      static void Main (string[] args) {
+         MyList<int> list = new ();
+         Console.WriteLine ($"Count before adding elements in the list: {list.Count}");
+         Console.WriteLine ($"Capacity before adding elements: {list.Capacity}");
+         list.Add (0);
+         list.Add (1);
+         list.Add (2);
```
Comment (priyarathna) [inline Training/Program.cs:105]: Including list.Remove (1) at line number 106, program gets crashed.

File: Training/Program.cs [hunk 10]
```diff
@@ -1,7 +1,131 @@
...
+      }
+
+      /// <summary>Removes the element at the specified index</summary>
+      /// <param name="index"></param>
+      /// <exception cref="IndexOutOfRangeException"></exception>
+      public void RemoveAt (int index) {
+         if (index < 0 || index >= mVariables.Length) throw new IndexOutOfRangeException ("Index is out of range");
+         for (int i = index; i < mVariables.Length - 1; i++)
+            mVariables[i] = mVariables[i + 1];
+         Array.Resize (ref mVariables, mVariables.Length - 1);
```
Comment (priyarathna) [inline Training/Program.cs:83]: What is the purpose of this?

File: Training/Program.cs [hunk 11]
```diff
@@ -1,7 +1,131 @@
...
+      public void Clear () => mSize = 0;
+
+      /// <summary>Insert elements at a specified index</summary>
+      /// <param name="index"></param>
+      /// <param name="element"></param>
+      /// <exception cref="IndexOutOfRangeException"></exception>
+      public void Insert (int index, T element) {
+         if (index < 0 || index >= mVariables.Length) throw new IndexOutOfRangeException ("Index is out of range");
+         if (Count == Capacity) Array.Resize (ref mVariables, Capacity * 2);
+         mVariables[mSize++] = element;
```
Comment (priyarathna) [inline Training/Program.cs:71]: I think this is not needed. Please check.

**Review by priyarathna — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 10 above)
Comment (Shresht07) [reply Training/Program.cs:83]: To shift the elements to left side, for example if 0th element is removed, then the 1st element must be shifted to 0th place.

File: Training/Program.cs [hunk 12]
```diff
@@ -1,7 +1,133 @@
...
+         list.Add (6);
+         list.Add (7);
+         PrintMyList (list);
+         Console.WriteLine ($"Count after adding elements in the list: {list.Count}");
+         Console.WriteLine ($"Capacity after adding elements: {list.Capacity}");
+         list.RemoveAt (0);
+         list.Insert (1, 3);
+         PrintMyList (list);
+         list.Clear ();
+         Console.Write ("After clearing list, list count => ");
```
Comment (priyarathna) [inline Training/Program.cs:119]: Can write it in a single line   Console.Write ($"After clearing list, list count => {list.Count}");

File: Training/Program.cs [hunk 13]
```diff
@@ -1,7 +1,133 @@
...
+         Console.WriteLine (list.Count);
+
+         /// <summary>Prints the elements of list in certain order</summary>
+         /// <param name="list"></param>
+         static void PrintMyList (MyList<int> list) {
```
Comment (priyarathna) [inline Training/Program.cs:124]: Include 'PrintList' in List class itself.

**Review by priyarathna — CHANGES_REQUESTED** [review]

**Review by priyarathna — APPROVED** [review]

**Pull request closed without merging**